### PR TITLE
Add mondido gateway

### PR DESCRIPTION
**Pull request by anderson-mondido**
Hello,

We're from Mondido Payments (mondido.com), a Swedish company.
We want to provide Shopify with an Payment provider for the Scandinavian market integrating our gateway into Active Merchant gem.

To perform the remote tests, we may provide some credentials privately or you can register an account at Mondido.com and set up your merchant_id, api_token, password and RSA public_key (need all of them).

We also created a [documentation for MondidoGateway](https://github.com/anderson-mondido/active-merchant-mondido/wiki). Note: this documentation URL may change in the future.

Matt of our team will send an email to "payment-integrations@shopify.com" to formalize this PR.
Please feel free to ask us any questions that may arise.

The Mondido Team

### CODE REVIEW

File: lib/active_merchant/billing/gateways/mondido.rb [hunk 1]
```diff
@@ -0,0 +1,658 @@
...
+      }
+
+      def initialize(options={})
+        requires!(options, :merchant_id, :api_token, :hash_secret)
+
+        @merchant_id = options[:merchant_id]
+        @api_token = options[:api_token]
+        @hash_secret = options[:hash_secret]
+
+        # Optional: RSA Encryption
```
Comment (ntalbott) [inline lib/active_merchant/billing/gateways/mondido.rb]: Can we rip out all the "optional" stuff? AM strives for simple, cross-gateway implementations. If someone really wants to go "all in" on Mondido-specific security, they could implement directly against your API.

File: lib/active_merchant/billing/gateways/mondido.rb [hunk 2]
```diff
@@ -0,0 +1,658 @@
...
+          # test bool
+          #   Must be true if you are using a test card number.
+          :test => test?
+        }
+
+        # customer_ref  string
+        #   Merchant specific customer ID.
+        #   If this customer exists the card will be added to that customer.
+        #   If it doesn't exists a customer will be created.
+        post.merge!( :customer_ref => options[:customer_ref].to_s ) if options[:customer_ref]
```
Comment (ntalbott) [inline lib/active_merchant/billing/gateways/mondido.rb]: Can we simplify to just `:customer_id` instead of needing both? `:customer_ref` is not AM standard, which means it breaks the AM abstraction.

Comment (ntalbott) [inline lib/active_merchant/billing/gateways/mondido.rb]: Whoops, I mean just `:customer`, as that's the standard AM option for specifying a customer profile identifier.

File: lib/active_merchant/billing/gateways/mondido.rb [hunk 3]
```diff
@@ -0,0 +1,658 @@
...
+        options[:order_id] = options[:order_id].to_s
+
+        post = {
+          # string* required
+          # The ID of the merchant
+          :merchant_id => @merchant_id,
+
+          # decimal* required
+          # The transaction amount ex. 12.00
+          :amount => get_amount(money, options),
```
Comment (ntalbott) [inline lib/active_merchant/billing/gateways/mondido.rb]: Methods prefixed with `get_` are a Ruby anti-pattern. Just use `amount`.

File: lib/active_merchant/billing/gateways/mondido.rb [hunk 4]
```diff
@@ -0,0 +1,658 @@
...
+          post.merge!( :metadata => options[:metadata] ) 
+        end
+
+        # Store Card (boolean)
+        #   true/false if you want to store the card
+        post.merge!( :store_card => options[:store_card] ) if options[:store_card]
+
+        # Plan ID (int)
+        #   The ID of the subscription plan.
+        post.merge!( :plan_id => options[:plan_id] ) if options[:plan_id]
```
Comment (ntalbott) [inline lib/active_merchant/billing/gateways/mondido.rb]: Please remove all functionality related to subscriptions/recurring. They're outside of scope for ActiveMerchant.

File: lib/active_merchant/billing/gateways/mondido.rb (hunk 1 above)
Comment (anderson-mondido) [inline lib/active_merchant/billing/gateways/mondido.rb]: Hi Nathaniel, yes we can rip out all the de "optional" stuff if that's what takes to integrate. I just want to point that it's security for the client, not the server. It solve the key distribution problem with some trade-off. Actually, as a user of ActiveMerchant, I would love to have this feature for all gateways. Now talking as Mondido, ok, we rip it out. Just a couple of minutes.

File: lib/active_merchant/billing/gateways/mondido.rb (hunk 2 above)
Comment (anderson-mondido) [inline lib/active_merchant/billing/gateways/mondido.rb]: Sure, will do now.

File: lib/active_merchant/billing/gateways/mondido.rb [hunk 5]
```diff
@@ -0,0 +1,658 @@
...
+      def add_credit_card(post, credit_card)
+        if credit_card.is_a? String      
+          post[:card_number] = credit_card
+          post[:card_type] = "STORED_CARD"
+        else
+          post[:card_holder] = credit_card.name if credit_card.name
+          post[:card_cvv] = credit_card.verification_value if credit_card.verification_value?
+          post[:card_expiry] = format(credit_card.month, :two_digits) + format(credit_card.year, :two_digits)
+          post[:card_number] = credit_card.number
+          post[:card_type] = ActiveMerchant::Billing::CreditCard.brand?(credit_card.number)
```
Comment (ntalbott) [inline lib/active_merchant/billing/gateways/mondido.rb]: Why not just `credit_card.brand`?

File: lib/active_merchant/billing/gateways/mondido.rb [hunk 6]
```diff
@@ -0,0 +1,658 @@
...
+      end
+
+      def add_support_params(post, options)
+        # extend  string
+        #   Extentability
+        #   Many objects have child resources such as customer and stored card.
+        #   To minimize the object graph being sent over REST calls you can choose to add full objects by
+        #   specifying to extend them in the url.
+        query_string_values = {}
+        query_string_values[:extend] = options[:extend].to_s if options[:extend]
```
Comment (ntalbott) [inline lib/active_merchant/billing/gateways/mondido.rb]: This seems really out of scope for AM. Remove.

File: lib/active_merchant/billing/gateways/mondido.rb [hunk 7]
```diff
@@ -0,0 +1,658 @@
...
+        #   The API supports responses in Swedish and English so if you specify locale=sv or locale=en either
+        #   as a parameter that you send in or int the query you will receive error messages in the desired language.
+        query_string_values[:locale] = options[:locale].to_s if options[:locale]
+
+        # start_id int
+        # limit    int
+        # offset   int
+        #   Partial data
+        #   While doing API calls you can use pagination to fetch parts of your data using limit, offset and start_id.
+        query_string_values[:start_id] = options[:start_id].to_s if options[:start_id]
```
Comment (ntalbott) [inline lib/active_merchant/billing/gateways/mondido.rb]: AM shouldn't be making any calls that need pagination. Remove.

File: lib/active_merchant/billing/gateways/mondido.rb [hunk 8]
```diff
@@ -0,0 +1,658 @@
...
+        query_string_values[:limit] = options[:limit].to_s if options[:limit]
+        query_string_values[:offset] = options[:offset].to_s if options[:offset]
+
+        if not query_string_values.empty?
+          post[:query_string_values] = query_string_values
```
Comment (ntalbott) [inline lib/active_merchant/billing/gateways/mondido.rb]: Remove.

File: lib/active_merchant/billing/gateways/mondido.rb [hunk 9]
```diff
@@ -0,0 +1,658 @@
...
+        # Query String
+        query_string = ""
+        if parameters.key?(:query_string_values)
+          query_string = "?" + URI.encode_www_form(parameters[:query_string_values])
+
+          parameters[:query_string_values] = nil
+          parameters.delete(:query_string_values)
+        end
+
+        if parameters.is_a?(Hash) and parameters.count == 0
```
Comment (ntalbott) [inline lib/active_merchant/billing/gateways/mondido.rb]: Please don't use `and`, always prefer `&&` in Ruby.

File: lib/active_merchant/billing/gateways/mondido.rb [hunk 10]
```diff
@@ -0,0 +1,658 @@
...
+          parameters = nil
+        end
+
+        # Flow for non pinned requests
+        # Make sure to use ssl_request because ActiveMerchant already handle
+        # Root CAs verification.
+        if not (@certificate_for_pinning or @certificate_hash_for_pinning or @public_key_for_pinning)
```
Comment (ntalbott) [inline lib/active_merchant/billing/gateways/mondido.rb]: Same with `not` vs. `!`, and `or` vs. `||`.

File: lib/active_merchant/billing/gateways/mondido.rb [hunk 11]
```diff
@@ -0,0 +1,658 @@
...
+                      OpenSSL::Digest::SHA256.hexdigest(parameters[:cert].to_pem)
+        end
+      end
+
+      def headers(options = {})
+        {
+          "Authorization" => "Basic " + Base64.encode64("#{@merchant_id}:#{@api_token}").strip,
+          "User-Agent" => "Mondido ActiveMerchantBindings/#{ActiveMerchant::VERSION}",
+          "Content-Type" => "application/x-www-form-urlencoded"
+          #"X-Mondido-Client-User-Agent" => user_agent, # defined in Gateway.rb
```
Comment (ntalbott) [inline lib/active_merchant/billing/gateways/mondido.rb]: Remove any commented out code.

File: test/remote/gateways/remote_mondido_test.rb [hunk 12]
```diff
@@ -0,0 +1,1487 @@
+require 'test_helper'
+
+class RemoteMondidoTest < Test::Unit::TestCase
```
Comment (ntalbott) [inline test/remote/gateways/remote_mondido_test.rb]: Overall this remote test looks way over-complicated; you shouldn't need all the support methods. Reference other gateway remote tests for how to use methods on the gateway (such as `store`) instead of re-implementing in the test.

File: lib/active_merchant/billing/gateways/mondido.rb (hunk 4 above)
Comment (anderson-mondido) [inline lib/active_merchant/billing/gateways/mondido.rb]: Ok, will do.

File: lib/active_merchant/billing/gateways/mondido.rb (hunk 3 above)
Comment (anderson-mondido) [inline lib/active_merchant/billing/gateways/mondido.rb]: Ok, will do.

**Comment by ntalbott** [comment]
@anderson-mondido looks like a good start, but needs a lot of simplification. If you can address my comments - and feel free to push back, happy to discuss any/all - then I'll give it another review.

File: lib/active_merchant/billing/gateways/mondido.rb (hunk 1 above)
Comment (ntalbott) [inline lib/active_merchant/billing/gateways/mondido.rb]: @anderson-mondido I agree certificate pinning would be great to have as a core ActiveMerchant feature, just doesn't make sense to do it on a gateway-by-gateway basis.

As for adding another layer of encryption inside of SSL, I'm not a fan - just adds complexity without IMO any real security gain.

File: lib/active_merchant/billing/gateways/mondido.rb (hunk 5 above)
Comment (anderson-mondido) [inline lib/active_merchant/billing/gateways/mondido.rb]: Good point. Probably saw in some gateway. Using .brand now.

File: lib/active_merchant/billing/gateways/mondido.rb (hunk 6 above)
Comment (anderson-mondido) [inline lib/active_merchant/billing/gateways/mondido.rb]: Removed

File: lib/active_merchant/billing/gateways/mondido.rb (hunk 7 above)
Comment (anderson-mondido) [inline lib/active_merchant/billing/gateways/mondido.rb]: Removed

File: lib/active_merchant/billing/gateways/mondido.rb (hunk 8 above)
Comment (anderson-mondido) [inline lib/active_merchant/billing/gateways/mondido.rb]: Removed

File: lib/active_merchant/billing/gateways/mondido.rb (hunk 9 above)
Comment (anderson-mondido) [inline lib/active_merchant/billing/gateways/mondido.rb]: Thanks. Learned something new today. Changes applied.

File: lib/active_merchant/billing/gateways/mondido.rb (hunk 10 above)
Comment (anderson-mondido) [inline lib/active_merchant/billing/gateways/mondido.rb]: Changes applied.

File: lib/active_merchant/billing/gateways/mondido.rb (hunk 11 above)
Comment (anderson-mondido) [inline lib/active_merchant/billing/gateways/mondido.rb]: Removed

File: test/remote/gateways/remote_mondido_test.rb (hunk 12 above)
Comment (anderson-mondido) [inline test/remote/gateways/remote_mondido_test.rb]: Was an attempt to run remote tests without merchant specific account. Changed to use hard coded values, but need to pass the credentials for you using a private channel.

Comment (ntalbott) [inline test/remote/gateways/remote_mondido_test.rb]: @anderson-mondido can we get a shared AM test account to check in? It makes ongoing maintenance *way* easier for us maintainers.

**Comment by anderson-mondido** [comment]
@ntalbott I've simplified as you said and applied all your requests. Please, could you give another review? Thank you very much :)

File: lib/active_merchant/billing/gateways/mondido.rb [hunk 13]
```diff
@@ -0,0 +1,448 @@
...
+      def scrub(transcript)
+        transcript.
+          gsub(%r((Authorization: Basic )\w+), '\1[FILTERED]').
+          gsub(%r((card_holder=)\w+), '\1[FILTERED]').
+          gsub(%r((card_cvv=)\d+), '\1[FILTERED]').
+          gsub(%r((card_expiry=)\w+), '\1[FILTERED]').
+          gsub(%r((card_number=)\d+), '\1[FILTERED]').
+          gsub(%r((card_type=)\w+), '\1[FILTERED]').
+          gsub(%r((hash=)\w+), '\1[FILTERED]').
+          gsub(%r((amount=)\w+), '\1[FILTERED]')
```
Comment (girasquid) [inline lib/active_merchant/billing/gateways/mondido.rb]: Why scrub the amount?

Comment (anderson-mondido) [inline lib/active_merchant/billing/gateways/mondido.rb]: I asked myself why not? It's not very sensistive as the other values. I can remove. Should I?

Comment (ntalbott) [inline lib/active_merchant/billing/gateways/mondido.rb]: So actually `scrub` is targeted primarily/totally at credentials and PCI-controlled data, so really `card_holder`, `card_expiry`, `card_type`, and `amount` should be left alone.

Comment (girasquid) [inline lib/active_merchant/billing/gateways/mondido.rb]: Yes - you should only scrub things that are sensitive. Amounts are useful for debugging and not sensitive.

**Comment by ntalbott** [comment]
We're in the middle of PCI L1 audit here at Spreedly - <sarcasm>yay!</sarcasm> - so I'm moving slowly, but I should be able to work through AM backlog in a couple weeks and get this merged.

File: lib/active_merchant/billing/gateways/mondido.rb (hunk 13 above)
Comment (anderson-mondido) [inline lib/active_merchant/billing/gateways/mondido.rb]: Alright! Removed from scrub :) - Thanks. I'm learning during this process hehe :D Seriously.

**Comment by anderson-mondido** [comment]
Oh good to hear that you're being audited. It's always good to test your controls :)
Even not beign able to merge this week, do you want to perform the remote tests in advance? I can send the credentials in private for you.

File: lib/active_merchant/billing/gateways/mondido.rb (hunk 1 above)
Comment (anderson-mondido) [inline lib/active_merchant/billing/gateways/mondido.rb]: Take Heartbleed vulnerability as an example. You'll be able to read the remote server's memory. What happens if the value that you read is encrypted? It's just defense in depth, but yes, you're right about the complexity. Plus the server may have its security requirements reduced as it will recept encrypted data. Depends on more factors, but that's good for my explanation I think.

**Comment by anderson-mondido** [comment]
@ntalbott Hope PCI audit has gone well. Do you happen to have any idea of when our gateway might be merged? Would be great to have a rough estimate to know how to prioritize the roadmap on our end.

**Comment by ntalbott** [comment]
Nope. Super busy here.

**Comment by anderson-mondido** [comment]
Ok, thanks. Please let know when you could.

File: test/remote/gateways/remote_mondido_test.rb (hunk 12 above)
Comment (j-mutter) [inline test/remote/gateways/remote_mondido_test.rb]: Any update on getting credentials for this that we can commit to the repo?

Comment (anderson-mondido) [inline test/remote/gateways/remote_mondido_test.rb]: Do you have a email that I can send privetly?

Comment (j-mutter) [inline test/remote/gateways/remote_mondido_test.rb]: So that's a 'no' for ones we can make public? I'll take them privately, but would love to have a set we can commit here and use in CI

Comment (anderson-mondido) [inline test/remote/gateways/remote_mondido_test.rb]: Some tests require some scenarios that are specific so I've used my account. Was dynamically before but I was asked to remove in one of the previous comments. You can take the acc that I send you privetely for your tests then. Can you send me an email to anderson {at} mondido.com?

**Comment by ntalbott** [comment]
@anderson-mondido I should be able to work on this soon, but need credentials. If we can't get public ones, please email me directly with them.

**Comment by robbanp** [comment]
@ntalbott I'll send you Id, password and secret

**Comment by ntalbott** [comment]
@anderson-mondido I got a chance to run the remote tests with the credentials you sent, and I'm seeing these failures:

```
  test_successful_authorize_stored_card:		F
  test_successful_authorize_stored_card_metadata:	F
  test_successful_authorize_stored_card_webhook:	F
  test_successful_authorize_stored_card_webhook_metadata:F
  test_successful_purchase_stored_card:			F
  test_successful_purchase_stored_card_metadata:	F
  test_successful_purchase_stored_card_webhook:		F
  test_successful_purchase_stored_card_webhook_metadata:F
  test_successful_void:					F
```

Further, the remote tests are actually **too* extensive - ActiveMerchant doesn't need or want to test every permutation of possible input. Please have a look at other gateway remote tests for an idea of how extensive the tests should be.

**Comment by robbanp** [comment]
@ntalbott, Thanks for the feedback. We'll update and resubmit!

**Comment by jasonwebster** [comment]
Closing due to staleness. Feel free to reopen when the outstanding issues are resolved.
